pkg/operator/ceph/testlib/spec.go

```diff
@@ -0,0 +1,67 @@
+
+// TestArguments returns a descriptive error if any of the expected arguments do not exist.
+// This supports arguments in which flags appear multiple times with different values but does not
+// support multiple instances of the same flag-value. This test is designed to fail if the list
+// of actual arguments contains extra arguments not specified in the expected args.
+// The expected arguments are given as an array of string arrays. This is to support flags which
+// may have multiple values. Examples:
+// expectedArgs := [][]string{
+//     {"-h"},                              // test for a short flag
+//     {"-vvv"},                            // test for a short flag with value(s) specified
+//     {"-d", "3"},                         // test for a short flag with a value specified
+//     {"--verbose"},                       // test for a --bool flag
+//     {"--name=alex"},                     // test for a --flag=value flag
+//     {"--name", "sam"},                   // test for a --flag with a value after a space
+//     {"--full-name", "sam", "goodhuman"}, // test for a --flag with 2 values separated by spaces
+// }
+func TestArguments(actualArgs []string, expectedArgs [][]string) error {
+	// join all args into a big space-separated arg string so we can use string search on it
+	// this is simpler than a bunch of nested for loops and if statements with continues in them
+	fullArgString := strings.Join(actualArgs, " ")
+	logger.Info("testing that actual args: %s\nmatch expected args:%v", fullArgString, actualArgs)
+	for _, arg := range expectedArgs {
+		// We join each individual argument together the same was as the big string
+		validArgMatcher := strings.Join(arg, " ")
+		if validArgMatcher == "" {
+			return fmt.Errorf("Expected argument %v evaluated to empty string; TestArguments() doesn't know what to do", arg)
+		}
+		matches := strings.Count(fullArgString, validArgMatcher)
+		if matches > 1 {
+			return fmt.Errorf("More than one instance of flag '%s' in: %s; TestArguments() doesn't know what to do",
+				validArgMatcher, fullArgString)
+		} else if matches == 1 {
+			// Remove the instance of the valid match so we can't match to it any more
+			fullArgString = strings.Replace(fullArgString, validArgMatcher, "", 1)
+		} else { // zero matches
+			return fmt.Errorf("Expected argument '%s' missing in: %s\n(It's possible the same arg is in expectedArgs twice.)",
+				validArgMatcher, strings.Join(actualArgs, " "))
+		}
+	}
+	if remainingArgs := strings.Trim(fullArgString, " "); remainingArgs != "" {
+		return fmt.Errorf("The actual arguments have additional args specified: %s", remainingArgs)
+	}
+	return nil
+}
+
+// AddCephCommonDaemonArgsToExpectedArgs appends all of the common Ceph daemon args to expected args
+func AddCephCommonDaemonArgsToExpectedArgs(namespace string, expectedArgs [][]string) [][]string {
+	return append(expectedArgs,
+		[]string{"--cluster", namespace},
+		[]string{"--conf", "/etc/ceph/ceph.conf"},
```

Comment:
one thing to confirm is that the data persisted by the mons needs to be under the dataDirHostPath. If we change the mon to use /etc/ceph, does that also put the persisted data there, or could we still point it to /var/lib/rook where the dataDirHostPath is mounted?

Reply:
I think the mon should still store data in `/var/lib/rook` for now. As I have thought about the config/data location work, that has a few different goals:
 - Ceph's config should live in `/etc/ceph` - this is `ceph.conf` and `keyring` - that's basically happening in the init refactor work here
 - Ceph's data should live in Ceph's default data location, which I believe is `/var/lib/ceph` - that's not happening here
 - `/var/lib/ceph` in the container would map to something like `/var/lib/rook/ceph/<namespace>/[mon|osd]/<daemon-name>` to prevent different clusters and different daemons from stomping on each others' data. - also not happening here

Having written all this down, I think it's safe to allow the config to change to `/etc/ceph/` since that makes sense for these PRs. For now, the mons still use `/var/lib/rook`, and the mgr does as well. With the future data location change, `/var/lib/ceph/mon-<name>/<data>` will persist to `/var/lib/ceph/<ns>/mon-<name>/<data>`. And for the mgr, `/var/lib/ceph/<mgr>-<name>/<data>` will not persist to disk (though there is an option to do that...) but will persist to an empty dir.

I've had a lot of time recently to think about the config/data path rework since it seems adjacent to this, so I may as well work on that after these init refactors are done. I'll see if I can assign myself to that issue.

Reply:
One challenge with changing the path of persistent data is that we need to support upgrades. If the path on the host changes, we would need to support both the old and new paths so it may not be worth changing. Where things map inside the container can change without a problem though.

Reply:
I was just considering this. I think it would be possible to have the config init container mount both the old and new locations for a transition period and copy from the old location to the new location.

Another thought was about Rook 1.0. Are we assuming that Rook 1.0 won't be able to upgrade from Rook 0.X (beta) and that breaking changes are okay? If that's the case, this work can wait until then.

Reply:
We are planning to support all upgrades going forward. If there are changes, we at least need to document how to migrate even if the upgrade is not fully automated.

If we can solve it with mounting tricks that would be fine, but I would not recommend we copy.